# <u>**Extract, Transform and Load (ETL)**</u>

This notebook is the intial step toward a standard data analytics workflow. Data quality is key to ensure effective Exploratory Data Analysis, Feature Engineering, Machine Learning Model development and dashboard visualisation. 

The data cleaning process requires thorough checks to ensure the integrity of the datasets are reviewed before saving transformations into a joined DataFrame. The steps taken in this notebook are:

- Load libraries and datasets

- Initial datasets review 

- Standardise column names 

- Identify missing, unique or duplicate values

- Ammend DataTypes 

- Add year columns for year-on-year analysis

- Join 2023-2024 and 2025-2025 datasets

- Drop identifiable information columns  

- Save DataFrame into clean data folder

ETL can reveal additional demands from the datasets during these checks. This plan is based on the most common steps taught by Code Institute LMS content, tutors and online resources included in the credits section of the README. Built-in Visual Studio Code Copilot will also collaboratively fix, modify and add comments to streamline the process.

</b>

## <u>**Load Libraries and Data**</u>

Libraries - Pandas and Numpy - are imported. Raw datasets exported as CSV from EXCEL sheets are loaded into two seperate DataFrames using Pandas.  

In [15]:
# Import libraries 
import pandas as pd
import numpy as np

# Read the CSV file into a DataFrame
df_24 = pd.read_csv('/Users/admin/Documents/Code Institute/ace_project/data/raw/2023_24_ace_raw.csv')
df_26 = pd.read_csv('/Users/admin/Documents/Code Institute/ace_project/data/raw/2025_26_ace_raw.csv')

# Check rows and columns in the DataFrames
print("Shape of df_24:", df_24.shape)
print("Shape of df_26:", df_26.shape)

Shape of df_24: (3085, 11)
Shape of df_26: (2369, 11)


</b>

### <u>**Intial DataFrame Review**</u>

Use <code>.info</code> and <code>.head()</code> for an intial review on both DataFrames.

In [16]:
df_24.info()  # Display information about df_24
df_24.head()  # Display the first few rows of df_24

<class 'pandas.DataFrame'>
RangeIndex: 3085 entries, 0 to 3084
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   Recipient              3085 non-null   str  
 1   Activity name          3085 non-null   str  
 2   Award amount           3085 non-null   str  
 3   Award date             3085 non-null   str  
 4   Decision Month         3085 non-null   str  
 5   Decision quarter       3085 non-null   str  
 6   ACE Area               3085 non-null   str  
 7   Local authority        3085 non-null   str  
 8   Main discipline        3085 non-null   str  
 9   Strand                 3085 non-null   str  
 10  Time-Limited Priority  240 non-null    str  
dtypes: str(11)
memory usage: 265.2 KB


,Recipient,Activity name,Award amount,Award date,Decision Month,Decision quarter,ACE Area,Local authority,Main discipline,Strand,Time-Limited Priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


In [17]:
df_26.info()  # Display information about df_26
df_26.head()  # Display the first few rows of df_26

<class 'pandas.DataFrame'>
RangeIndex: 2369 entries, 0 to 2368
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   Recipient              2369 non-null   str  
 1   Activity name          2369 non-null   str  
 2   Award amount           2369 non-null   str  
 3   Award date             2369 non-null   str  
 4   Decision Month         2369 non-null   str  
 5   Decision quarter       2369 non-null   str  
 6   ACE Area               2369 non-null   str  
 7   Local authority        2368 non-null   str  
 8   Main discipline        2369 non-null   str  
 9   Strand                 2369 non-null   str  
 10  Time-Limited Priority  224 non-null    str  
dtypes: str(11)
memory usage: 203.7 KB


,Recipient,Activity name,Award amount,Award date,Decision Month,Decision quarter,ACE Area,Local authority,Main discipline,Strand,Time-Limited Priority
0,Window Wanderland CIC,"Community of practise, Wanderland Pilot","£33,522",01-Apr-25,April,Q1,South West,"Bristol, City of",Combined arts,"£30,000 and below",NaN
1,Cat and Mouse Theatre,Time Crime,"£27,461",01-Apr-25,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN
2,London Borough of Ealing Libraries and communi...,Y-Our Story Visual Stories of Acton,"£27,000",01-Apr-25,April,Q1,London,Ealing,Libraries,"£30,000 and below",Universal Library Offers
3,Exmouth Festival,Exmouth Festival 2025,"£38,500",01-Apr-25,April,Q1,South West,East Devon,Combined arts,"Above £30,000",NaN
4,Crediton Heart Project,Crediton Heart Summer Arts Festival,"£15,135",01-Apr-25,April,Q1,South West,Mid Devon,Combined arts,"£30,000 and below",NaN


The outputs for <code>.info()</code> shows both DataFrames share the same titles for column names but will require text standardisation. 

Missing values are present in both DataFrames <code>df_24</code> shows 240 rows contain values in Time-Limited Priority column. Meanwhile, <code>df_26</code> has 224 values for Time-Limited Priority and one missing value in the Local Authority column. 

DataTypes are all set as strings, which require ammending into numerical, data-time and categorical types to plot visualisation in the EDA notebook.

Identifiable information is shown in the Recipient and Activity name from the output for <code>.head()</code>. These columns must be dropped to ensure ethical data handling. 

</b>

## <u>**Standardise Column Names**</u>

Format must be consistent to access columns in the EDA, Feature Engineering and ML Modeling notebooks. Remove leading or trailing whitespace use <code>.strip()</code>, replace spaces or hyphens with underscores use <code>.replace()</code> and convert text to lowercase use <code>.lower()</code>. These functions are a professional standard protocol for technical users to navigate the code cells.  

In [18]:
# Format column names
df_24.columns = (df_24.columns
                 .str.strip()
                 .str.lower()
                 .str.replace(' ', '_')
                 .str.replace('-', '_'))

df_26.columns = (df_26.columns
                 .str.strip()
                 .str.lower()
                 .str.replace(' ', '_')
                 .str.replace('-', '_'))


In [19]:
# Check column names 2023-2024
df_24.head() 

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time_limited_priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


In [20]:
# Check column names 2025-2026
df_24.head()

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time_limited_priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


</b>

## <u>**Missing Values**</u>

### **Inspect Missing Values**

Missing values can lower the accuracy of visualisations in the EDA and ML model performance. NaN or blank values can also cause errors for tools and libraries in the ML Modeling notebook. To inspect missing values use <code>.isna().sum()</code>.

In [21]:
# Check for missing values in df_24
df_24.isna().sum()

recipient                   0
activity_name               0
award_amount                0
award_date                  0
decision_month              0
decision_quarter            0
ace_area                    0
local_authority             0
main_discipline             0
strand                      0
time_limited_priority    2845
dtype: int64

Time-Limited priority has 2,854 missing values for awards received between 2023-2024. Further research to establish the purpose of this column can help determine how to handle these missing values.

In [22]:
# Check for missing values in df_26
df_26.isna().sum()  

recipient                   0
activity_name               0
award_amount                0
award_date                  0
decision_month              0
decision_quarter            0
ace_area                    0
local_authority             1
main_discipline             0
strand                      0
time_limited_priority    2145
dtype: int64

Again, the time-limited priority column flags 2,145 missing values that need to be considered prior to transformation. This DataFrame for awards received between 2025-2026 also contains one missing value in the Local Authorities column.

In [23]:
df_26[df_26["local_authority"].isna()]

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time_limited_priority
1328,Comics Cultural Impact Collective,NLPG O30 Oct25,"£37,000",17-Dec-25,December,Q3,North,NaN,Literature,"Above £30,000",NaN


According to official Arts Council documentation:

>We define place as a specific geographic area of any scale (village, town, local authority, city or city region) where the proposed intervention will take place and where a cultural strategy (or similar) exists that covers the same geography. This could mean that an application might be for activity in a town or village that is covered by a cultural strategy for a wider district.

[National Lottery Project Grants Information Sheet](https://www.artscouncil.org.uk/sites/default/files/2023-09/Place%20Partnership%20projects%20and%20Project%20Grants%20-%20Information%20sheet.pdf)

The Arts Council website also addresses the Time-limited priority column. The missing values represent projects that are not time-limited. 

[National Lottery Project Grants: Guidance Library](https://www.artscouncil.org.uk/ProjectGrants/national-lottery-project-grants-guidance-library#t-in-page-nav-5)

</b>

### **Handle Missing Values**

Replace missing value in the 2025-2026 DataFrame for <code>local_authority</code> with "Unknown", as this recipient potentially operates in more than one local authority according to the Arts Council documentation. 

In [24]:
# Replace missing value in the 'local_authority' column with 'Unknown'
df_26['local_authority'] = df_26['local_authority'].fillna('Unknown')

# Check for remaining missing values in df_26
df_26.isna().sum()

recipient                   0
activity_name               0
award_amount                0
award_date                  0
decision_month              0
decision_quarter            0
ace_area                    0
local_authority             0
main_discipline             0
strand                      0
time_limited_priority    2145
dtype: int64

Replace missing values in both DataFrames 2023-2024 and 2025-2026 for the <code>time_limited_priority</code> column with "Not Time Limited Priority". Arts Council documentation confirms time-limited priority does not apply to all awards.

In [25]:
# Replace missing values in the 'time_limited_priority' columns
for df in [df_24, df_26]:
   df['time_limited_priority'] = (
       df['time_limited_priority']
       .fillna('Not Time-Limited Priority'))

# Check for remaining missing values in df_24 and df_26
print("Missing values in df_24 after replacement:")
print(df_24.isna().sum())
print("\nMissing values in df_26 after replacement:")
print(df_26.isna().sum())


Missing values in df_24 after replacement:
recipient                0
activity_name            0
award_amount             0
award_date               0
decision_month           0
decision_quarter         0
ace_area                 0
local_authority          0
main_discipline          0
strand                   0
time_limited_priority    0
dtype: int64

Missing values in df_26 after replacement:
recipient                0
activity_name            0
award_amount             0
award_date               0
decision_month           0
decision_quarter         0
ace_area                 0
local_authority          0
main_discipline          0
strand                   0
time_limited_priority    0
dtype: int64


</b>

## **<u>Duplicate Rows</u>**

Check for duplicate rows use <code>.duplicated().sum()</code> to identify and sum up awards that appear more than once. This step improves accuracy in the EDA and ML Modeling notebooks.

In [26]:
# Check for duplicates in df_24 and df_26
print("Duplicates in df_24:", df_24.duplicated().sum())
print("Duplicates in df_26:", df_26.duplicated().sum())

Duplicates in df_24: 0
Duplicates in df_26: 0


The output displays no duplicate rows. Thus, no further action is required to clean the DataFrames.

<b>

## **<u>Unique Values Review</u>**

Review unique values in <code>ace_area</code>, <code>local_authority</code>, <code>main_discipline</code>, <code>strand</code> and <code>time_limited_priority</code> columns use <code>.unique()</code> to confirm consistency. A single category may be represented by multiple categories due to inconsistent spelling, spaces and capitalisation. As a result, these values may distort group counts and filtering in visualisations. 

In [28]:
""" ACE AREA """

# Check unique values in ace_area column in df_24
print("Unique values in 'ace_area' column of df_24:")
print(df_24['ace_area'].unique())    

# Check unique values in ace_area column in df_26
print("Unique values in 'ace_area' column of df_26:")
print(df_26['ace_area'].unique()) 

Unique values in 'ace_area' column of df_24:
<StringArray>
['South West', 'Midlands', 'London', 'South East', 'North']
Length: 5, dtype: str
Unique values in 'ace_area' column of df_26:
<StringArray>
['South West', 'London', 'Midlands', 'South East', 'North']
Length: 5, dtype: str


In [29]:
""" LOCAL AUTHORITY """

# Check unique values in the 'local_authority' column of df_26
print("Unique values in 'local_authority' column of df_26:")
print(df_26['local_authority'].unique())    

# Check unique values in the 'local_authority' column of df_24
print("Unique values in 'local_authority' column of df_24:")
print(df_24['local_authority'].unique())    

Unique values in 'local_authority' column of df_26:
<StringArray>
[            'Bristol, City of',        'South Gloucestershire',
                       'Ealing',                   'East Devon',
                    'Mid Devon',                   'Wandsworth',
                    'Islington',                      'Hackney',
 'Bath and North East Somerset',                     'Tamworth',
 ...
                       'Melton',                    'Hertsmere',
                    'Wokingham',                      'Fenland',
                        'Powys',             'Scottish Borders',
                     'Stirling',                     'Havering',
                     'Bolsover',          'South Staffordshire']
Length: 282, dtype: str
Unique values in 'local_authority' column of df_24:
<StringArray>
['Bournemouth, Christchurch and Poole',                          'Portsmouth',
        'Bath and North East Somerset',                    'Bristol, City of',
               'South Glouceste

In [38]:
""" STRAND """

# Check unique values in the 'strand' column of df_26
print("Unique values in 'strand' column of df_26:")
print(df_26['strand'].unique())    

# Check unique values in the 'strand' column of df_24
print("Unique values in 'strand' column of df_24:")
print(df_24['strand'].unique())    

Unique values in 'strand' column of df_26:
<StringArray>
[                        '£30,000 and below',
                             'Above £30,000',
             'Above £30,000 - Major Project',
 'Above £30,000 - Over £30k Touring project',
    'Above £30,000 - Over £100k application',
 'Above £30,000 - Place Partnership project']
Length: 6, dtype: str
Unique values in 'strand' column of df_24:
<StringArray>
[                             '£30,000 and below',
                                  'Above £30,000',
         'Above £30,000 - Over £100k application',
      'Above £30,000 - Over £30k Touring project',
      'Above £30,000 - Place Partnership project',
 'Above £30,000 - Nationally Significant project']
Length: 6, dtype: str


In [36]:
""" TIME LIMITED PRIORITY """

# Check unique values in the 'time_limited_priority' column of df_26
print("Unique values in 'time_limited_priority' column of df_26:")
print(df_26['time_limited_priority'].unique())    

# Check unique values in the 'time_limited_priority' column of df_24
print("Unique values in 'time_limited_priority' column of df_24:")
print(df_24['time_limited_priority'].unique())   

Unique values in 'time_limited_priority' column of df_26:
<StringArray>
[               'Not Time-Limited Priority',
                 'Universal Library Offers',
                    'Unlocking Collections',
              'Supporting Grassroots Music',
 'Research and Development for Individuals']
Length: 5, dtype: str
Unique values in 'time_limited_priority' column of df_24:
<StringArray>
['Supporting Grassroots Live Music',        'Not Time-Limited Priority',
            'Unlocking Collections',      'Supporting Grassroots Music',
         'Universal Library Offers']
Length: 5, dtype: str


### **Summary**

The output for each column displays the values do not require text standardisation. However, there are difference between the two 2023-2024 and 2025-2026 in the <code>local_authority</code>, <code>strand</code> and <code>time_limited_priority</code> coumns. 

These differences are to be expected since official documentation by Arts Council communicates their strategy is subject to change each year for <code>strand</code> and <code>time_limited_priority</code>. 

The <code>local_authority</code> column is very noisy due to the granular nature of local authorities. However, they could still be grouped to count awards distributed across local authorities. 

</b>

## **<u>Convert DataTypes</u>**

Data consistency is neccessary to plot visualisations in the EDA notebook. Review DataTypes and convert each column to numerical, date-time or categorical. Use </code><code>.dtypes()</code> to display the current DataTypes in each DateFrame. 

The DataFrames contain mostly categorical columns, one numeric value and date-time column which were identified as strings in the intial review of both DataFrames.

In [39]:
# Check DataTypes for 2023-2024 and 2025-2026
print("Data types for df_24:")
print(df_24.dtypes)

print("\nData types for df_26:")
print(df_26.dtypes)

Data types for df_24:
recipient                str
activity_name            str
award_amount             str
award_date               str
decision_month           str
decision_quarter         str
ace_area                 str
local_authority          str
main_discipline          str
strand                   str
time_limited_priority    str
dtype: object

Data types for df_26:
recipient                str
activity_name            str
award_amount             str
award_date               str
decision_month           str
decision_quarter         str
ace_area                 str
local_authority          str
main_discipline          str
strand                   str
time_limited_priority    str
dtype: object


</b >

### **Award Amount**

Similar to standardising column names in the previous section <code>.replace()</code> and <code>.strip()</code> remove symbols and hidden spaces to avoid ValueError caused by an invalid literal. Convert string into numeric DataType use <code>.astype(int)</code>.

In [40]:
# Clean and convert the 'award_amount' column in df_24
df_24["award_amount"] = (df_24["award_amount"]
                         .str.replace(',','')
                         .str.replace('£','')
                         .str.strip()
                         .astype(int))


# Clean and convert the 'award_amount' column in df_26
df_26["award_amount"] = (df_26["award_amount"]
                         .str.replace(',','')
                         .str.replace('£','')
                         .str.strip()
                         .astype(int))


# Check the data type of the 'award_amount' column
print("Data type df_24:", df_24["award_amount"].dtype) 
print("Data type df_26:", df_26["award_amount"].dtype) 


Data type df_24: int64
Data type df_26: int64


</b>

### **Date-Time**

Convert award-date column from string into date-time. Use <code>pd.to_datetime</code> for both DateFrames and check the conversion is successful. 

Add a column dedicated to year and improve filtering or grouping rows. Use <code>.dt.year</code> to extract the year from award date column.  

In [45]:
# Convert the award_date column to datetime for both DataFrames
for data in (df_24, df_26):
    data['award_date'] = pd.to_datetime(
        data['award_date'],
        errors='coerce'
    )

    # Extract the award year from the datetime values
    data['award_year'] = (
        data['award_date']
        .dt.year
        .astype('Int64')
    )

Standardise the date format and handles any invalid dates safely by coercing them to NaT (missing values) instead of raising an error use <code>error='coerce'</code>.

In [43]:
# Check the first few rows of df_24 to verify the changes
df_24.head()   

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time_limited_priority,award_year
0,Feelprime Ltd,Old Fire Station Digital,27689,2023-04-03,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music,2023
1,Downtown Pompey,Sticky Floors and Palm Trees,29623,2023-04-03,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",Not Time-Limited Priority,2023
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,25256,2023-04-03,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",Not Time-Limited Priority,2023
3,Upfest,Weston Wallz 2023,30500,2023-04-03,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",Not Time-Limited Priority,2023
4,Chloe Thurlow,"Haunt, Pray, Love",20060,2023-04-03,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",Not Time-Limited Priority,2023


Intially, the built-in copilot modified the code to remove time-stamp in the column with <code>.normalize()</code> which did not solve the issue because its function is to reset the time to midnight. Instead, <code>.dt.date</code> was the appropiate method to remove the time-stamp from column.

However, this causes an obstacle to extract the year to create a column in the next step. Thus, prior to removing the time-stamp add column then remove time-stamp.

In [46]:
# Remove time from award_date column in df_24 and df_26
df_24['award_date'] = df_24['award_date'].dt.date
df_26['award_date'] = df_26['award_date'].dt.date

In [47]:
# Check the first few rows of df_24 to verify the changes
df_24.head()  

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time_limited_priority,award_year
0,Feelprime Ltd,Old Fire Station Digital,27689,2023-04-03,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music,2023
1,Downtown Pompey,Sticky Floors and Palm Trees,29623,2023-04-03,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",Not Time-Limited Priority,2023
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,25256,2023-04-03,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",Not Time-Limited Priority,2023
3,Upfest,Weston Wallz 2023,30500,2023-04-03,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",Not Time-Limited Priority,2023
4,Chloe Thurlow,"Haunt, Pray, Love",20060,2023-04-03,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",Not Time-Limited Priority,2023


In [48]:
# Check the first few rows of df_26 to verify the changes
df_26.head()

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time_limited_priority,award_year
0,Window Wanderland CIC,"Community of practise, Wanderland Pilot",33522,2025-04-01,April,Q1,South West,"Bristol, City of",Combined arts,"£30,000 and below",Not Time-Limited Priority,2025
1,Cat and Mouse Theatre,Time Crime,27461,2025-04-01,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",Not Time-Limited Priority,2025
2,London Borough of Ealing Libraries and communi...,Y-Our Story Visual Stories of Acton,27000,2025-04-01,April,Q1,London,Ealing,Libraries,"£30,000 and below",Universal Library Offers,2025
3,Exmouth Festival,Exmouth Festival 2025,38500,2025-04-01,April,Q1,South West,East Devon,Combined arts,"Above £30,000",Not Time-Limited Priority,2025
4,Crediton Heart Project,Crediton Heart Summer Arts Festival,15135,2025-04-01,April,Q1,South West,Mid Devon,Combined arts,"£30,000 and below",Not Time-Limited Priority,2025


</b>

### **Categorical**

Categorical data types contain a list of unique values that represent different categories. Use <code>.astype("category")</code> to convert strings into a categorical data type. 

In [49]:
# Group categorical columns for conversion to 'category' data type
categorical_cols = [
    "main_discipline",
    "strand",
    "time_limited_priority",
    "ace_area"
]

# Convert the categorical columns
for df in [df_24, df_26]:
    for col in categorical_cols:
        if col in df.columns:
            df[col] = df[col].astype("category")


In [ ]:
# Check the data types of the categorical columns in df_24
df_24.dtypes

recipient                     str
activity_name                 str
award_amount                int64
award_date                 object
decision_month                str
decision_quarter              str
ace_area                 category
local_authority               str
main_discipline          category
strand                   category
time_limited_priority    category
award_year                  Int64
dtype: object

In [ ]:
# Check the data types of the categorical columns in df_24
df_24.dtypes

# Check the data types of the categorical columns in df_26
df_26.dtypes

recipient                     str
activity_name                 str
award_amount                int64
award_date                 object
decision_month                str
decision_quarter              str
ace_area                 category
local_authority               str
main_discipline          category
strand                   category
time_limited_priority    category
award_year                  Int64
dtype: object

The <code>time_limited_priority</code>column remained as a string. In the intial review missing values were identified which could be impacting the coversion. Thus, missing, duplicate and unique values are assessed prior to converting data types. 

Whilst handling missing values in <code>time_limited_priority</code> a hyphen was found and replaced with an underscore in the *Standardise Column Names* section. This highlights the importance of standardising column name titles at the beginning of the ETL to avoid errors.

</b>

## **<u>Anonymisation</u>**

Remove <code>recipient</code> and <code>activity_name</code> columns because they contain identifiable information and are not required for EDA, Feature Engineering, ML Model notebooks or dashboard visualisations use <code>.drop()</code> to drop the columns. This step ensures the dataset respects ethical data handling and focus on relevant features. 

In [52]:
# Remove identifiers from both DataFrames
df_24 = df_24.drop(columns=['recipient', 'activity_name'])
df_26 = df_26.drop(columns=['recipient', 'activity_name'])

In [54]:
# Check columns after dropping identifiers
print("Columns in df_24 after dropping identifiers:", df_24.columns)
print("Columns in df_26 after dropping identifiers:", df_26.columns)    

Columns in df_24 after dropping identifiers: Index(['award_amount', 'award_date', 'decision_month', 'decision_quarter',
       'ace_area', 'local_authority', 'main_discipline', 'strand',
       'time_limited_priority', 'award_year'],
      dtype='str')
Columns in df_26 after dropping identifiers: Index(['award_amount', 'award_date', 'decision_month', 'decision_quarter',
       'ace_area', 'local_authority', 'main_discipline', 'strand',
       'time_limited_priority', 'award_year'],
      dtype='str')


</b>

## **<u>Join DataFrames</u>**

Join clean 2023-2024 and 2025-2026 National Lottery Project Grants datasets into a single DataFrame and reset the index use <code>pd.concat()</code> and <code>ignore_index=True</code>

In [58]:
# Concatenate the two DataFrames into a single DataFrame
df_combined = pd.concat([df_24, df_26], ignore_index=True)

# Check the shape of the combined DataFrame
print("Shape of combined DataFrame:", df_combined.shape)

# Check the first few rows of the combined DataFrame
df_combined.head()

Shape of combined DataFrame: (5454, 10)


,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time_limited_priority,award_year
0,27689,2023-04-03,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music,2023
1,29623,2023-04-03,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",Not Time-Limited Priority,2023
2,25256,2023-04-03,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",Not Time-Limited Priority,2023
3,30500,2023-04-03,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",Not Time-Limited Priority,2023
4,20060,2023-04-03,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",Not Time-Limited Priority,2023


</b>

## **<u>Save Clean DataFrame</u>**

The final step is to save the <code>df_combined</code> DataFrame for the upcoming EDA and Feature Engineering notebooks use <code>.to_csv</code> and title the clean DataFrame appropriately.

In [59]:
# Save the cleaned DataFrame as a new CSV file
df.to_csv("cleaned_data.csv", index=False)